packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -102,28 +105,78 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 		const maxSizeBytes = maxSize && !isNaN(maxSize) ? maxSize * 1000 : 0;
 		const formattedMaxFileSize = formatFileSize(maxSizeBytes);
 
-		const [fileRejections, setFileRejections] = useState<RejectedFile[]>([]);
+		let validMaxFiles = maxFiles;
 
-		const handleRemoveFile = (file: FileWithStatus) => {
-			const indexOfFile = value.indexOf(file);
-			onChange(value.filter((_, index) => index !== indexOfFile));
-		};
+		if (maxFiles !== undefined && maxFiles < 1) {
+			validMaxFiles = undefined;
 
-		const handleRemoveRejection = (fileName: string) => {
-			setFileRejections(
-				fileRejections.filter((err) => err.file.name !== fileName)
+			console.warn(
+				'FileUpload: maxFiles cannot be less than 1. The property is being ignored.'
 			);
+		}
+
+		const [tooManyFilesRejections, setTooManyFilesRejections] = useState<
+			FileRejection[]
+		>([]);
+		const [invalidRejections, setInvalidRejections] = useState<FileRejection[]>(
+			[]
+		);
+		const [allRejections, setAllRejections] = useState<FileRejection[]>([]);
+
+		useEffect(() => {
+			setAllRejections([...invalidRejections, ...tooManyFilesRejections]);
+		}, [invalidRejections, tooManyFilesRejections]);
```

Comment:
Could this just be: 

```ts
const allRejections = [...invalidRejections, ...tooManyFilesRejections];
```

Reply:
It could, but it's expensive to spread those errors on every re-render while you're dragging

Reply:
Oh right, this is what we spoke about yesterday, I think. Maybe memoizing is better because this creates a second re-render.

Reply:
Done